logs/querysample/querysample.go

```diff
@@ -0,0 +1,406 @@
+			normalizeExprArray(&groupKey) // value comes from deparse_expression
+			(*groupingSet.GroupKeys)[idx] = groupKey
+		}
+	}
+	if groupingSet.HashKeys != nil {
+		for idx, hashKey := range *groupingSet.HashKeys {
+			normalizeExprArray(&hashKey) // value comes from deparse_expression
+			(*groupingSet.HashKeys)[idx] = hashKey
+		}
+	}
+	normalizeExprArray(groupingSet.SortKey) // value comes from deparse_expression
+}
+
+func normalizePlanNodeFields(node planNode) {
+	normalizeExpr(node.CacheKey)       // value comes from deparse_expression
+	normalizeExpr(node.ConflictFilter) // value comes from show_upper_qual
+	normalizeExpr(node.Filter)         // value comes from show_scan_qual
+	normalizeExpr(node.FunctionCall)   // value comes from show_expression
+	normalizeExprArray(node.GroupKey)  // value comes from deparse_expression
+	if node.GroupingSets != nil {
+		for _, groupingSet := range *node.GroupingSets {
+			normalizeGroupingSet(groupingSet)
+		}
+	}
+	normalizeExpr(node.HashCond)                // value comes from show_upper_qual
+	normalizeExpr(node.IndexCond)               // value comes from show_scan_qual
+	normalizeExpr(node.JoinFilter)              // value comes from show_upper_qual
+	normalizeExpr(node.MergeCond)               // value comes from show_upper_qual
+	normalizeExpr(node.OneTimeFilter)           // value comes from show_upper_qual
+	normalizeExpr(node.OrderBy)                 // value comes from show_scan_qual
+	normalizeExprArray(node.Output)             // value comes from deparse_expression
+	normalizeExprArray(node.PresortedKey)       // value comes from deparse_expression
+	normalizeExpr(node.RecheckCond)             // value comes from show_scan_qual
+	normalizeExpr(node.RepeatableSeed)          // value comes from deparse_expression
+	normalizeExprArray(node.SamplingParameters) // value comes from deparse_expression
+	if node.SortKey != nil {
+		for idx, expr := range *node.SortKey {
+			normalizeSortKey(&expr) // value comes from deparse_expression (but may have sort order information)
+			(*node.SortKey)[idx] = expr
+		}
+	}
+	normalizeExpr(node.TableFunctionCall) // value comes from show_expression
+	normalizeExpr(node.TIDCond)           // value comes from show_scan_qual
+
+	if node.Workers != nil {
+		for _, p := range *node.Workers {
+			normalizePlanNodeFields(p)
+		}
+	}
+
+	for _, p := range node.Plans {
+		normalizePlanNodeFields(p)
+	}
+}
+
+func normalizePlan(planRaw json.RawMessage) (json.RawMessage, error) {
+	var plan planNode
+	var err error
+	if err = json.Unmarshal(planRaw, &plan); err != nil {
+		return planRaw, err
+	}
+	normalizePlanNodeFields(plan)
+	return json.Marshal(plan)
+}
+
+func transformExplainJSONToQuerySample(logLine state.LogLine, explainText string, queryRuntimeMs float64, normalize bool) (state.PostgresQuerySample, error) {
+	var planDetails PlanContainer
+
+	if err := json.Unmarshal([]byte(explainText), &planDetails); err != nil {
```

Comment:
Does this mean we're parsing and re-formatting the JSON even if we're not normalizing? We're not doing this today, are we? Do we need to do that? I don't know if this is still an issue, but I believe the `encoding/json` performance was [not great](https://github.com/json-iterator/go#benchmark). Could this be a performance issue for users with large plans?

Reply:
Yeah, we already do this - in fact this approach is now slightly improving on the performance of it, because we no longer unmarshal the actual plan (we keep that as json.RawMessage) unless the `normalize` setting is active.

In my understanding there are two reasons we want to do this:

* Avoid generating samples that have broken JSON (due to truncated messages)
* Remove the `Query Text` field from the top level node, since its duplicative of what we already remember separately

We could consider dropping this step (since we already parse the JSON on the server-side anyway). I'll look into doing this together with refactoring where we do the normalization.

Reply:
Ah. After some more digging, we can't actually avoid this, because the query text itself is only contained within the JSON. We need to parse the JSON so we can get the query, so we can then fingerprint it (and associate the query sample and log line). I'll thus keep the JSON parsing.